### ISSUE TITLE
DateTime parsing on WebAssembly broken after trimming in latest .NET 7 main

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/runtime#69891]
### Description

After trimming, simple date parsing calls throw.

### Reproduction Steps

Add the following to code running on WebAssembly with the current runtime/SDK:

```cs
var parsed = DateTime.Parse("6/27/2022 11:08:05 AM");
```

... then trim the application (e.g., if it's Blazor, do `dotnet publish -c Release` and run that output).

### Expected behavior

Should parse the date

### Actual behavior

```
dotnet.7.0.0-preview.5.22274.8.ekt40wn8gu.js:3 
        
       Error: [MONO] * Assertion: should not be reached at /__w/1/s/src/mono/mono/mini/interp/interp.c:3606

    at mono_wasm_stringify_as_error_with_stack (dotnet.7.0.0-preview.5.22274.8.ekt40wn8gu.js:3:19927)
    at Object.mono_wasm_trace_logger (dotnet.7.0.0-preview.5.22274.8.ekt40wn8gu.js:3:20298)
    at _mono_wasm_trace_logger (dotnet.7.0.0-preview.5.22274.8.ekt40wn8gu.js:12:105856)
    at dotnet.wasm:0x9eb2
    at dotnet.wasm:0xa21ed
    at dotnet.wasm:0x1bed4d
    at dotnet.wasm:0x1beddc
    at dotnet.wasm:0x1bee45
    at dotnet.wasm:0xd5fd
    at dotnet.wasm:0xcab0
```

### Regression?

Yes, worked in earlier .NET 7 previews and of course in .NET 6 and earlier.

### Known Workarounds

None

### Other information

This appears to be the underlying cause for the Blazor E2E tests failure that started happening exactly when we updated the SDK in https://github.com/dotnet/aspnetcore/pull/41158

**I know this might be an SDK/trimmer issue, but I'm reporting it in runtime because it also might be that the runtime needs to react to trimmer changes and do something to ensure the culture-aware parsing mechanisms survive trimming.** In any case it's probably the same people who will fix this regardless of the repo, so please feel free to move the issue to SDK if applicable.

This is clearly a ship-blocking issue for .NET 7. It would be valuable to resolve this in time for the next preview release otherwise WebAssembly customers won't really be able to use that preview for its intended preview purpose. Thanks!

**Comment by lewing** [comment]
cc @vitek-karas

**Comment by tarekgh** [comment]
CC @marek-safar

**Comment by radical** [comment]
TL;dr It was linker issue which is already fixed in newer versions, and in newer sdks.

More info:
- I was able to reproduce this with `7.0.100-preview.5.22263.22`, which I got from https://github.com/dotnet/aspnetcore/blob/main/global.json#L3 .
- can also reproduce this if `DateTime.Parse(..)` is called from `Main`
- but I *cannot* reproduce this with newer versions

After lot of digging with @lewing, we found that it was a linker bug which is fixed in newer versions. For example, if I override the linker version to `7.0.100-1.22270.1`(which is from newer sdks), then it works fine:

```xml
    <PackageReference Include="Microsoft.NET.ILLink.Tasks"
                      Version="7.0.100-1.22270.1"
                      PrivateAssets="all"
                      IsImplicitlyDefined="true" />
```

Closing the issue.
